modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/CatalogValidationSchemasSource.java

```diff
@@ -0,0 +1,192 @@
+        );
+    }
+
+    @Override
+    public List<FullTableSchema> tableSchemaVersionsBetween(int tableId, HybridTimestamp fromIncluding, int toIncluding) {
+        int fromCatalogVersion = catalogService.activeCatalogVersion(fromIncluding.longValue());
+
+        return catalogVersionToTableVersionSpansCache.computeIfAbsent(
+                new CatalogVersionToTableVersionSpan(tableId, fromCatalogVersion, toIncluding),
+                key -> tableSchemaVersionsBetweenCatalogAndTableVersions(tableId, fromCatalogVersion, toIncluding)
+        );
+    }
+
+    private List<FullTableSchema> tableSchemaVersionsBetweenCatalogVersions(int tableId, int fromCatalogVersion, int toCatalogVersion) {
+        return tableVersionsBetween(tableId, fromCatalogVersion, toCatalogVersion)
+                .map(CatalogValidationSchemasSource::fullSchemaFromTableDescriptor)
+                .collect(toList());
+    }
+
+    // It's ok to use Stream as the results of the methods that call this are cached.
+    private Stream<CatalogTableDescriptor> tableVersionsBetween(
+            int tableId,
+            int fromCatalogVersionIncluding,
+            int toCatalogVersionIncluding
+    ) {
+        return IntStream.rangeClosed(fromCatalogVersionIncluding, toCatalogVersionIncluding)
+                .mapToObj(catalogVersion -> catalogService.table(tableId, catalogVersion))
+                .filter(new Predicate<>() {
+                    int prevVersion = Integer.MIN_VALUE;
+
+                    @Override
+                    public boolean test(CatalogTableDescriptor tableDescriptor) {
+                        if (tableDescriptor.tableVersion() == prevVersion) {
+                            return false;
+                        }
+
+                        assert prevVersion == Integer.MIN_VALUE || tableDescriptor.tableVersion() == prevVersion + 1;
+
+                        prevVersion = tableDescriptor.tableVersion();
+
+                        return true;
+                    }
+                });
+    }
+
+    private List<FullTableSchema> tableSchemaVersionsBetweenCatalogAndTableVersions(
+            int tableId,
+            int fromCatalogVersion,
+            int toTableVersion
+    ) {
+        return tableVersionsBetween(tableId, fromCatalogVersion, catalogService.latestCatalogVersion())
+                .takeWhile(tableDescriptor -> tableDescriptor.tableVersion() <= toTableVersion)
+                .map(CatalogValidationSchemasSource::fullSchemaFromTableDescriptor)
+                .collect(toList());
+    }
+
+    private static FullTableSchema fullSchemaFromTableDescriptor(CatalogTableDescriptor tableDescriptor) {
+        return new FullTableSchema(
+                tableDescriptor.tableVersion(),
+                tableDescriptor.id(),
+                tableDescriptor.columns()
+        );
+    }
+
+    private static class CatalogVersionsSpan {
```

Comment:
Please move to separete package private class.

Reply:
These classes are very specific to this class, they are not needed by anyone else. Also, the class is not too long, so there is no urgency in moving them out.

I would prefer to leave them here.

Reply:
ok